libvmi/driver/kvm/kvm.c

```diff
@@ -101,6 +105,373 @@ translate_msr_index(int index, int *err) {
+    libvmi_event->x86_regs->ss_sel = kvmi_event->event.common.sregs.ss.selector;
+    //          ES
+    libvmi_event->x86_regs->es_base = kvmi_event->event.common.sregs.es.base;
+    libvmi_event->x86_regs->es_limit = kvmi_event->event.common.sregs.es.limit;
+    libvmi_event->x86_regs->es_sel = kvmi_event->event.common.sregs.es.selector;
+    //          FS
+    libvmi_event->x86_regs->fs_base = kvmi_event->event.common.sregs.fs.base;
+    libvmi_event->x86_regs->fs_limit = kvmi_event->event.common.sregs.fs.limit;
+    libvmi_event->x86_regs->fs_sel = kvmi_event->event.common.sregs.fs.selector;
+    //          GS
+    libvmi_event->x86_regs->gs_base = kvmi_event->event.common.sregs.gs.base;
+    libvmi_event->x86_regs->gs_limit = kvmi_event->event.common.sregs.gs.limit;
+    libvmi_event->x86_regs->gs_sel = kvmi_event->event.common.sregs.gs.selector;
+    //      VCPU
+    libvmi_event->vcpu_id = kvmi_event->event.common.vcpu;
+}
+
+static status_t
+process_cb_response(
+        vmi_instance_t vmi,
+        event_response_t response,
+        vmi_event_t *libvmi_event,
+        struct kvmi_dom_event *kvmi_event,
+        void *rpl,
+        size_t rpl_size)
+{
+    kvm_instance_t *kvm = kvm_get_instance(vmi);
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!kvm || !kvm->kvmi_dom) {
+        errprint("%s: invalid kvm or kvmi handles\n", __func__);
+        return VMI_FAILURE;
+    }
+    if (!libvmi_event || !kvmi_event || !rpl) {
+        errprint("%s: invalid libvmi/kvmi/rpl handles\n", __func__);
+        return VMI_FAILURE;
+    }
+#endif
+    // loop over all possible responses
+    for (uint32_t i = VMI_EVENT_RESPONSE_NONE+1; i <=__VMI_EVENT_RESPONSE_MAX; i++) {
+        event_response_t candidate = 1u << i;
+        if (response & candidate) {
+            switch (candidate) {
+                default:
+                    errprint("%s: KVM - unhandled event reponse %u\n", __func__, candidate);
+                    break;
+            }
+        }
+    }
+    struct kvmi_event_reply *_rpl = rpl;
+    _rpl->action = KVMI_EVENT_ACTION_CONTINUE;
+    _rpl->event = kvmi_event->event.common.event;
+
+    if (kvmi_reply_event(kvm->kvmi_dom, kvmi_event->seq, rpl, rpl_size))
+        return VMI_FAILURE;
+
+    return VMI_SUCCESS;
+}
+
+static event_response_t
+call_event_callback(
+        vmi_instance_t vmi,
+        vmi_event_t *libvmi_event)
+{
+    event_response_t response;
+    vmi->event_callback = 1;
+    response = libvmi_event->callback(vmi, libvmi_event);
+    vmi->event_callback = 0;
+    return response;
+}
+
+/*
+ * VM event handlers (process_xxx)
+ * called from kvm_events_listen
+ */
+static status_t
+process_register(vmi_instance_t vmi, struct kvmi_dom_event *event)
+{
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!vmi || !event)
+        return VMI_FAILURE;
+#endif
+    dbprint(VMI_DEBUG_KVM, "--Received CR event\n");
+    return VMI_SUCCESS;
+}
+
+static status_t
+process_msr(vmi_instance_t vmi, struct kvmi_dom_event *kvmi_event)
```

Comment:
@tklengyel a naïve implementation of an event handler that I choose:

- fill the libvmi event with `fill_ev_common_kvmi_to_libvmi()`
- call the user callback with `call_event_callback()`
- declare the kvmi reply struct
- process the callback response with `process_cb_response()`

As you have written the Xen driver, you may have suggestions how the code should be structured, which helpers should be used, etc.